### ISSUE TITLE
Optimizing `dart run` with incremental compilation

### ISSUE THREAD

**Issue by jakemac53** [dart-lang/sdk#47372]
One of the biggest complaints about Dart (since Dart 2 at least) is how long it takes to run normal scripts. The vm always starts over completely from scratch on each run of a command, so for scripts that import lots of deps, it can be 10s of seconds to compile the app to kernel, and in the meantime all you get is a blank screen.

Recently, I updated the `pub` tool to use the frontend server and incremental compilation even when running from source. This gives me a time sometimes as low as 1 second to compile apps where the changes are mostly to top level packages. A significant upgrade!

This leaves us in an interesting situation where `pub run <file.dart>` is actually significantly more performant than `dart <file.dart>` for subsequent builds.... but almost nobody knows this or would choose that option today :D.

So my question is, what would be the best way to bring this same support to the top level `dart` command?

## Idea 1 - Just use pub run 

Have `dart <file.dart>` (and `dart run <file.dart>`) actually just be aliases for `pub run`, if we detect that you are running from a package root (so `pub run` will succeed).

## Idea 2 - Duplicate the logic (maybe with some shared libs)

If we want the logic to work more consistently, regardless of use case, we could essentially just duplicate the logic that pub has today, but add support for running with no package config. Possibly we could even just move it to a shared library (or maybe it already is shared).

The tool would have its own directory for the cached kernel files though.

## Idea 3 - Move the logic to the `dart` command

For this idea we would just move all logic for running all types of executables out of pub and into the top level `dart` command.

Then `pub run` would only ever just turn around and invoke `dart run` directly. This possibly gets weird with global packages though.

cc @sigurdm @jonasfj @mit-mit @kevmoo @natebosch for comment

**Comment by jonasfj** [comment]
Afaik, `dart run` already calls into pub (as a library), I'm pretty sure it's now effectively the same as `pub run` -- ie. it already has these performance improvements.

**Comment by jakemac53** [comment]
> Afaik, `dart run` already calls into pub (as a library), I'm pretty sure it's now effectively the same as `pub run` -- ie. it already has these performance improvements.

It looks like that isn't the case for passing a file name directly - only if you are running a package executable?

**Comment by kevmoo** [comment]
Better, more consistent perf SGTM

**Comment by sigurdm** [comment]
> It looks like that isn't the case for passing a file name directly - only if you are running a package executable?

That is true - but also for `pub run` - they basically share the same logic. [^1].

I think the snapshotting/incremental precompilation logic could be moved to dartdev - pub could focus on package resolution and finding the dart file to execute. 

One question is: where we would store the snapshot/kernel-file for a `dart [run] foo.dart` invocation.

Right now `dart foo.dart` and more importantly `dart foo.snapshot` is a shortcut in that it never invokes dartdev. I think we rely on that (as a faster way to get exectuting) for examples in the binstubs for `pub global activate`.

[^1]: We have a hack here: https://github.com/dart-lang/sdk/blob/main/pkg/dartdev/lib/src/commands/run.dart#L236-L241 but also pub goes around the precompilation logic if passed just a dart file: https://github.com/dart-lang/pub/blob/master/lib/src/executable.dart#L286.

**Comment by jakemac53** [comment]
> That is true - but also for `pub run` - they basically share the same logic. [1](#user-content-fn-1-9d7f1d75456cfeb45d4a901023d7a8db).

It definitely works today for `pub run <file.dart>` 🤷‍♂️ . If you try running the same program with `dart pub run` versus `dart run` it will compile significantly faster (you can repro this by just importing analyzer into a dummy app and running it both ways).

**Comment by jakemac53** [comment]
Well, note that I didn't explicitly do that (I just passed `bin/test.dart`). Pub just seems to have automatically done that conversion. I actually noticed that it will do the same thing when the file is not under `bin` which is actually probably a bug 🙃

**Comment by ykmnkmi** [comment]
little tool to build binary on changes, [gist](https://gist.github.com/ykmnkmi/02f4810ecc6e501b2939bf9adfefa116)
